keps/sig-network/20200127-disable-service-type-lb-nodeports.md

```diff
@@ -0,0 +1,117 @@
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+As of today, a Kubernetes Service with Type=LoadBalancer will always allocate a node port. Though most implementations of Service Type=LoadBalancer do require node ports, there are several implementations that do not. This KEP proposes to add a new flag on the apiserver `--disable-loadbalancer-node-ports` to optionally disable node port allocation for Service Type=LoadBalancer.
+
+
+## Motivation
+
+There are several implementations of Service Type=LoadBalancer API today that do not consume the node ports automatically allocated by Kubernetes. Some examples include but are not limited to:
+* [MetalLB](https://github.com/danderson/metallb)
+* [kube-router](https://github.com/cloudnativelabs/kube-router)
+
+The drawbacks of allocating unused node ports are that:
+* the number of load balancers is now limited to the number of available node ports
+* node ports are exposed for an LB even though they are not used -- unecessarily exposed ports can fail regulatory/compliance requirements
+
+For clusters that have integrations for Service Type=LoadBalancer but don't require node ports should have the option to disable node port allocation.
+
+### Goals
+
+* disable node port for Service Type=LoadBalancer
+* don't change any existing default behavior for Service Type=LoadBalancer
+
+## Proposal
+
+A flag `--disable-loadbalancer-node-ports` will be added to the apiserver to disable node port allocations for Service Type=LoadBalancer. The flag's default value is `false` to ensure existing behavior is preserved.
+
+If `--disable-loadbalancer-node-ports` is `true`,  node port allocation/release for Service Type=LoadBalancer is skipped for all rest operations. If a Service Type=LoadBalancer already exists prior to the flag being set, the apiserver will still handle releasing its node port on delete.
+
+## Design Details
+
+### Test Plan
+
+The addition of this feature should not break any existing E2E / conformance tests since the default behavior has not changed.
+
+### Graduation Criteria
+
+This enhancement will be GA and on by default immediately since it does not change any default behaviors in Kubernetes. Significant testing/validation will be done to ensure that node port allocation is only disabled if specified on the apiserver.
+
+### Upgrade / Downgrade Strategy
+
+On upgrade, a user _must_ enable the `--disable-loadbalancer-node-ports` flag to use the new feature. Any existing Services with node port allocated will not be affected if a user disables loadbalancer node ports.
+
+On downgrade, the apiserver will exit since the `--disable-loadbalancer-node-ports` flag will not exist. Node port allocation for load balancers being re-enabled on downgrade should be expected by users since action is required to remove the original flag from the apiserver.
+
+### Version Skew Strategy
+
+Not applicable because only the apiserver will be updated for this feature.
```

Comment:
This needs at least some test cases and auditing in kube-proxy to make sure there are no bad assumptions.

Reply:
makes sense, will update